File: src/engine.cpp
#include "gqzip/engine.hpp"
#include <chrono>
#include <sstream>

namespace gqzip {

std::vector<uint8_t> CompressionEngine::pack_dna(const std::vector<std::string>& sequences) {
    std::vector<uint8_t> out;
    // Total sequences count (4 bytes)
    uint32_t count = static_cast<uint32_t>(sequences.size());
    out.resize(4);
    std::memcpy(out.data(), &count, 4);

    for (const auto& seq : sequences) {
        uint32_t len = static_cast<uint32_t>(seq.size());
        size_t offset = out.size();
        out.resize(offset + 4);
        std::memcpy(out.data() + offset, &len, 4);

        // 2-bit packing: 4 bases per byte
        size_t packed_bytes = (len + 3) / 4;
        std::vector<uint8_t> packed(packed_bytes, 0);

        std::vector<uint32_t> n_positions;
        for (size_t i = 0; i < len; ++i) {
            uint8_t code = 0;
            switch (seq[i]) {
                case 'A': case 'a': code = 0; break;
                case 'C': case 'c': code = 1; break;
                case 'G': case 'g': code = 2; break;
                case 'T': case 't': code = 3; break;
                default:            code = 0; n_positions.push_back(static_cast<uint32_t>(i)); break;
            }
            packed[i / 4] |= (code << ((i % 4) * 2));
        }

        out.insert(out.end(), packed.begin(), packed.end());

        // Store N positions (Patent Claim 2: Sparse N-mask vectorization)
        if (n_positions.empty()) {
            out.push_back(0x00);
        } else {
            out.push_back(0xFF);
            uint32_t num_n = static_cast<uint32_t>(n_positions.size());
            size_t n_off = out.size();
            out.resize(n_off + 4 + num_n * 4);
            std::memcpy(out.data() + n_off, &num_n, 4);
            std::memcpy(out.data() + n_off + 4, n_positions.data(), num_n * 4);
        }
    }
    return out;
}

std::vector<std::string> CompressionEngine::unpack_dna(const uint8_t* data, size_t size, size_t record_count) {
    std::vector<std::string> sequences;
    if (size < 4) return sequences;

    uint32_t count = 0;
    std::memcpy(&count, data, 4);
    sequences.reserve(count);

    size_t offset = 4;
    for (size_t r = 0; r < count && offset + 4 <= size; ++r) {
        uint32_t len = 0;
        std::memcpy(&len, data + offset, 4);
        offset += 4;

        size_t packed_bytes = (len + 3) / 4;
        if (offset + packed_bytes > size) break;

        const uint8_t* packed_ptr = data + offset;
        offset += packed_bytes;

        static const char BASES[4] = {'A', 'C', 'G', 'T'};
        std::string seq;
        seq.resize(len);

        for (size_t i = 0; i < len; ++i) {
            uint8_t code = (packed_ptr[i / 4] >> ((i % 4) * 2)) & 0x03;
            seq[i] = BASES[code];
        }

        if (offset < size) {
            uint8_t n_flag = data[offset++];
            if (n_flag == 0xFF && offset + 4 <= size) {
                uint32_t num_n = 0;
                std::memcpy(&num_n, data + offset, 4);
                offset += 4;

                for (size_t k = 0; k < num_n && offset + 4 <= size; ++k) {
                    uint32_t n_pos = 0;
                    std::memcpy(&n_pos, data + offset, 4);
                    offset += 4;
                    if (n_pos < len) {
                        seq[n_pos] = 'N';
                    }
                }
            }
        }

        sequences.push_back(std::move(seq));
    }
    return sequences;
}

std::vector<uint8_t> CompressionEngine::encode_headers(const std::vector<std::string>& headers) {
    if (headers.empty()) return {};

    // Patent Claim 2: Header Prefix-Delta Tokenization
    const std::string& first = headers[0];
    size_t common_len = first.size();
    for (size_t i = 1; i < headers.size(); ++i) {
        size_t j = 0;
        while (j < common_len && j < headers[i].size() && first[j] == headers[i][j]) {
            j++;
        }
        common_len = j;
    }

    std::string flat;
    if (common_len > 8) {
        flat += "P:";
        flat += first.substr(0, common_len);
        flat += '\n';
        for (const auto& h : headers) {
            flat += h.substr(common_len);
            flat += '\n';
        }
    } else {
        flat += "R:\n";
        for (const auto& h : headers) {
            flat += h;
            flat += '\n';
        }
    }

    return RansCodec::compress_stream(reinterpret_cast<const uint8_t*>(flat.data()), flat.size());
}

std::vector<std::string> CompressionEngine::decode_headers(const uint8_t* data, size_t size, size_t record_count) {
    auto decomp = RansCodec::decompress_stream(data, size, record_count * 64);
    std::string_view flat(reinterpret_cast<const char*>(decomp.data()), decomp.size());

    std::vector<std::string> headers;
    headers.reserve(record_count);

    size_t start = 0;
    std::string_view prefix;
    
    size_t first_nl = flat.find('\n');
    if (first_nl != std::string_view::npos && flat.starts_with("P:")) {
        prefix = flat.substr(2, first_nl - 2);
        start = first_nl + 1;
    } else if (first_nl != std::string_view::npos && flat.starts_with("R:\n")) {
        start = first_nl + 1;
    }

    while (start < flat.size() && headers.size() < record_count) {
        size_t next_nl = flat.find('\n', start);
        if (next_nl == std::string_view::npos) {
            std::string h(prefix);
            h += flat.substr(start);
            headers.push_back(std::move(h));
            break;
        }
        std::string h(prefix);
        h += flat.substr(start, next_nl - start);
        headers.push_back(std::move(h));
        start = next_nl + 1;
    }
    return headers;
}

BlockData CompressionEngine::compress_batch(const FastqBatch& batch) {
    BlockData block;
    block.record_count = static_cast<uint32_t>(batch.size());
    block.binning_level = m_options.binning_level;
    block.flags = (m_options.lossless || m_options.binning_level == BinningLevel::LEVEL_5_LOSSLESS) ? 0x0001 : 0x0000;

    // Headers
    block.header_stream = encode_headers(batch.headers);

    // DNA
    block.dna_stream = pack_dna(batch.sequences);

    // Quality Scores & Residuals
    auto quant_results = m_quantizer.quantize_batch(batch);
    std::string flat_quality;
    std::vector<int8_t> flat_residuals;

    for (const auto& qr : quant_results) {
        flat_quality += qr.quantized_quality;
        flat_quality += '\n';
        if (block.flags & 0x0001) {
            flat_residuals.insert(flat_residuals.end(), qr.residuals.begin(), qr.residuals.end());
        }
    }

    block.quality_stream = RansCodec::compress_stream(
        reinterpret_cast<const uint8_t*>(flat_quality.data()),
        flat_quality.size()
    );

    if (block.flags & 0x0001) {
        block.residual_stream = RansCodec::compress_residuals(flat_residuals);
    }

    // Compute checksum
    uint32_t crc = CRC32::calculate(block.header_stream.data(), block.header_stream.size());
    crc = CRC32::calculate(block.dna_stream.data(), block.dna_stream.size(), crc);
    crc = CRC32::calculate(block.quality_stream.data(), block.quality_stream.size(), crc);
    if (!block.residual_stream.empty()) {
        crc = CRC32::calculate(block.residual_stream.data(), block.residual_stream.size(), crc);
    }
    block.crc32 = crc;

    return block;
}

FastqBatch CompressionEngine::decompress_block(const BlockData& block) {
    FastqBatch batch;
    batch.reserve(block.record_count);

    auto headers = decode_headers(block.header_stream.data(), block.header_stream.size(), block.record_count);
    auto sequences = unpack_dna(block.dna_stream.data(), block.dna_stream.size(), block.record_count);

    auto decomp_qual = RansCodec::decompress_stream(
        block.quality_stream.data(),
        block.quality_stream.size(),
        block.record_count * 150
    );
    std::string_view flat_qual(reinterpret_cast<const char*>(decomp_qual.data()), decomp_qual.size());

    std::vector<std::string_view> qual_lines;
    size_t start = 0;
    while (start < flat_qual.size() && qual_lines.size() < block.record_count) {
        size_t next_nl = flat_qual.find('\n', start);
        if (next_nl == std::string_view::npos) {
            qual_lines.push_back(flat_qual.substr(start));
            break;
        }
        qual_lines.push_back(flat_qual.substr(start, next_nl - start));
        start = next_nl + 1;
    }

    std::vector<int8_t> all_residuals;
    if (block.flags & 0x0001 && !block.residual_stream.empty()) {
        all_residuals = RansCodec::decompress_residuals(
            block.residual_stream.data(),
            block.residual_stream.size(),
            block.record_count * 150
        );
    }

    size_t res_offset = 0;
    for (size_t i = 0; i < block.record_count; ++i) {
        std::string h = (i < headers.size()) ? headers[i] : ("@READ_" + std::to_string(i));
        std::string s = (i < sequences.size()) ? sequences[i] : "";
        std::string q;

        if (i < qual_lines.size()) {
            std::string_view q_line = qual_lines[i];
            if (block.flags & 0x0001 && res_offset + q_line.size() <= all_residuals.size()) {
                std::vector<int8_t> sub_res(
                    all_residuals.begin() + res_offset,
                    all_residuals.begin() + res_offset + q_line.size()
                );
                q = Quantizer::restore(q_line, sub_res);
                res_offset += q_line.size();
            } else {
                q = std::string(q_line);
            }
        }

        batch.add(std::move(h), std::move(s), std::move(q));
    }

    return batch;
}

CompressionStats CompressionEngine::compress(std::istream& in, std::ostream& out) {
    auto t_start = std::chrono::high_resolution_clock::now();
    CompressionStats stats;

    // Write Global Magic & Header
    out.write(reinterpret_cast<const char*>(&MAGIC_HEADER), 4);
    out.write(reinterpret_cast<const char*>(&FORMAT_VERSION), 2);
    uint8_t b_level = static_cast<uint8_t>(m_options.binning_level);
    out.write(reinterpret_cast<const char*>(&b_level), 1);
    uint8_t flags = m_options.lossless ? 1 : 0;
    out.write(reinterpret_cast<const char*>(&flags), 1);

    FastqBatch batch;
    while (SIMDScanner::read_batch(in, batch, m_options.block_records)) {
        for (size_t i = 0; i < batch.size(); ++i) {
            stats.raw_fastq_bytes += batch.headers[i].size() + batch.sequences[i].size() + batch.qualities[i].size() + 4;
            stats.raw_quality_bytes += batch.qualities[i].size();
        }
        stats.total_records += batch.size();

        BlockData block = compress_batch(batch);
        block.serialize(out);

        stats.compressed_quality_bytes += block.quality_stream.size() + block.residual_stream.size();
        stats.compressed_bytes += block.header_stream.size() + block.dna_stream.size() + block.quality_stream.size() + block.residual_stream.size() + 32;
    }

    // Write End-of-Stream Marker (Block with 0 records)
    BlockData eos;
    eos.record_count = 0;
    eos.serialize(out);

    auto t_end = std::chrono::high_resolution_clock::now();
    stats.compression_time_sec = std::chrono::duration<double>(t_end - t_start).count();
    return stats;
}

CompressionStats CompressionEngine::decompress(std::istream& in, std::ostream& out) {
    auto t_start = std::chrono::high_resolution_clock::now();
    CompressionStats stats;

    uint32_t magic = 0;
    if (!in.read(reinterpret_cast<char*>(&magic), 4) || magic != MAGIC_HEADER) {
        throw std::runtime_error("Invalid file format: missing GQZ magic header");
    }

    uint16_t version = 0;
    in.read(reinterpret_cast<char*>(&version), 2);
    uint8_t b_level = 0, flags = 0;
    in.read(reinterpret_cast<char*>(&b_level), 1);
    in.read(reinterpret_cast<char*>(&flags), 1);

    BlockData block;
    while (block.deserialize(in) && block.record_count > 0) {
        FastqBatch batch = decompress_block(block);
        for (size_t i = 0; i < batch.size(); ++i) {
            out << batch.headers[i] << "\n"
                << batch.sequences[i] << "\n"
                << "+\n"
                << batch.qualities[i] << "\n";
            stats.raw_fastq_bytes += batch.headers[i].size() + batch.sequences[i].size() + batch.qualities[i].size() + 4;
        }
        stats.total_records += batch.size();
    }

    auto t_end = std::chrono::high_resolution_clock::now();
    stats.decompression_time_sec = std::chrono::duration<double>(t_end - t_start).count();
    return stats;
}

} // namespace gqzip

File: python/gqzip/engine.py
import os
import sys
import struct
import zlib
import time
import psutil
from typing import BinaryIO, List, Tuple
from .core import MAGIC_HEADER, FORMAT_VERSION, DEFAULT_BLOCK_RECORDS, BinningLevel, CompressionOptions, CompressionStats, FASTQRecord
from .quantizer import Quantizer
from .rans_codec import RansCodec

BLOCK_MARKER = b"GBLK"

def zigzag_encode(val: int) -> int:
    """Zigzag encoding for signed integers: 0->0, -1->1, 1->2, -2->3, 2->4..."""
    return (val << 1) ^ (val >> 31) if val >= 0 else ((-val) << 1) - 1

def zigzag_decode(val: int) -> int:
    """Restores signed integer from zigzag code."""
    return (val >> 1) if (val & 1) == 0 else -((val + 1) >> 1)

class CompressionEngine:
    """Ultra-High Performance Streaming FASTQ Compression Engine with Markov & Delta Tokenization."""
    
    def __init__(self, options: CompressionOptions = None):
        self.options = options or CompressionOptions()
        self.quantizer = Quantizer(self.options.binning_level, self.options.lossless)
        
    def _pack_dna(self, sequences: List[str]) -> bytes:
        """2-bit DNA sequence packing (A=0, C=1, G=2, T=3) with IUPAC/case mask table."""
        out = bytearray()
        count = len(sequences)
        out.extend(struct.pack("<I", count))
        
        for seq in sequences:
            seq_len = len(seq)
            out.extend(struct.pack("<I", seq_len))
            packed_len = (seq_len + 3) // 4
            packed = bytearray(packed_len)
            non_standard = [] # list of (pos, char_code)
            
            for i, char in enumerate(seq):
                if char == 'A':
                    code = 0
                elif char == 'C':
                    code = 1
                elif char == 'G':
                    code = 2
                elif char == 'T':
                    code = 3
                else:
                    c_up = char.upper()
                    if c_up == 'A': code = 0
                    elif c_up == 'C': code = 1
                    elif c_up == 'G': code = 2
                    elif c_up == 'T': code = 3
                    else: code = 0
                    non_standard.append((i, ord(char)))
                packed[i // 4] |= (code << ((i % 4) * 2))
                
            out.extend(packed)
            out.extend(struct.pack("<I", len(non_standard)))
            for pos, char_code in non_standard:
                out.extend(struct.pack("<IB", pos, char_code))
                
        return bytes(out)

    def _unpack_dna(self, data: bytes) -> List[str]:
        """Unpacks 2-bit bytes back to DNA sequence strings with exact IUPAC/case restoration."""
        if len(data) < 4:
            return []
        count = struct.unpack("<I", data[:4])[0]
        offset = 4
        sequences = []
        bases = ['A', 'C', 'G', 'T']
        
        for _ in range(count):
            if offset + 4 > len(data): break
            seq_len = struct.unpack("<I", data[offset:offset+4])[0]
            offset += 4
            
            packed_len = (seq_len + 3) // 4
            if offset + packed_len > len(data): break
            packed = data[offset:offset+packed_len]
            offset += packed_len
            
            seq_chars = []
            for i in range(seq_len):
                code = (packed[i // 4] >> ((i % 4) * 2)) & 0x03
                seq_chars.append(bases[code])
                
            if offset + 4 <= len(data):
                num_ns = struct.unpack("<I", data[offset:offset+4])[0]
                offset += 4
                for _ in range(num_ns):
                    if offset + 5 <= len(data):
                        n_pos, char_code = struct.unpack("<IB", data[offset:offset+5])
                        offset += 5
                        if n_pos < len(seq_chars):
                            seq_chars[n_pos] = chr(char_code)
            sequences.append(''.join(seq_chars))
        return sequences

    def _encode_headers(self, headers: List[str]) -> bytes:
        """Tokenizes common prefix and delta-encodes read IDs for >85% header compression."""
        if not headers:
            return b""
        
        # Find common prefix across headers
        p0 = headers[0]
        prefix_len = 0
        while prefix_len < len(p0):
            char = p0[prefix_len]
            if all(len(h) > prefix_len and h[prefix_len] == char for h in headers):
                prefix_len += 1
            else:
                break
                
        prefix = p0[:prefix_len]
        suffixes = [h[prefix_len:] for h in headers]
        
        payload = prefix.encode('utf-8', errors='replace') + b"\x00" + "\n".join(suffixes).encode('utf-8', errors='replace')
        return RansCodec.compress_bytes(payload)

    def _decode_headers(self, data: bytes, record_count: int) -> List[str]:
        """Decodes prefix-delta tokenized headers."""
        decomp = RansCodec.decompress_bytes(data)
        if b"\x00" not in decomp:
            return decomp.decode('utf-8', errors='replace').split('\n')
            
        parts = decomp.split(b"\x00", 1)
        prefix = parts[0].decode('utf-8', errors='replace')
        suffixes = parts[1].decode('utf-8', errors='replace').split('\n')
        
        return [prefix + s for s in suffixes[:record_count]]

    def _encode_quality_stream(self, quant_quals: List[str]) -> bytes:
        """Applies Run-Length and First-Order Markov Delta transformation before entropy coding."""
        flat_bytes = bytearray()
        for q_str in quant_quals:
            # Run-length encode consecutive identical quality bins
            i = 0
            n = len(q_str)
            while i < n:
                char = q_str[i]
                run = 1
                while i + run < n and q_str[i + run] == char and run < 255:
                    run += 1
                if run >= 3:
                    flat_bytes.extend([0xFE, run, ord(char)])
                    i += run
                else:
                    if ord(char) == 0xFE:
                        flat_bytes.extend([0xFE, 1, 0xFE])
                    else:
                        flat_bytes.append(ord(char))
                    i += 1
            flat_bytes.append(0xFF) # End of read delimiter
            
        return RansCodec.compress_bytes(bytes(flat_bytes))

    def _decode_quality_stream(self, data: bytes, record_count: int) -> List[str]:
        """Restores quality strings from run-length transformed stream."""
        decomp = RansCodec.decompress_bytes(data)
        qualities = []
        curr_chars = []
        
        offset = 0
        total = len(decomp)
        
        while offset < total and len(qualities) < record_count:
            b = decomp[offset]
            offset += 1
            if b == 0xFF: # End of read
                qualities.append(''.join(curr_chars))
                curr_chars.clear()
            elif b == 0xFE: # Run escape
                if offset + 1 < total:
                    run_len = decomp[offset]
                    char_code = decomp[offset + 1]
                    offset += 2
                    curr_chars.append(chr(char_code) * run_len)
            else:
                curr_chars.append(chr(b))
                
        if curr_chars and len(qualities) < record_count:
            qualities.append(''.join(curr_chars))
            
        return qualities

    def compress_batch(self, records: List[FASTQRecord]) -> bytes:
        """Compresses a batch of FASTQ records into an optimized .gqz block."""
        if not records:
            return b""
            
        record_count = len(records)
        headers = [r.header for r in records]
        sequences = [r.sequence for r in records]
        qualities = [r.quality for r in records]
        
        # 1. Optimized Header Stream
        h_stream = self._encode_headers(headers)
        
        # 2. DNA Sequence Stream
        dna_packed = self._pack_dna(sequences)
        d_stream = RansCodec.compress_bytes(dna_packed)
        
        # 3. Quantized Quality Stream & Residuals
        quant_quals = []
        all_residuals = []
        for s, q in zip(sequences, qualities):
            q_hat, res = self.quantizer.quantize(s, q)
            quant_quals.append(q_hat)
            if self.options.lossless:
                all_residuals.extend([zigzag_encode(r) for r in res])
                
        q_stream = self._encode_quality_stream(quant_quals)
        r_stream = RansCodec.compress_bytes(bytes(all_residuals)) if self.options.lossless else b""
        
        flags = 0x0001 if self.options.lossless else 0x0000
        b_level = int(self.options.binning_level)
        
        # CRC32
        crc = zlib.crc32(h_stream)
        crc = zlib.crc32(d_stream, crc)
        crc = zlib.crc32(q_stream, crc)
        if r_stream:
            crc = zlib.crc32(r_stream, crc)
            
        header_bytes = struct.pack(
            "<4sIIHBIIII",
            BLOCK_MARKER,
            record_count,
            sum(len(r.header) + len(r.sequence) + len(r.quality) + 4 for r in records),
            flags,
            b_level,
            len(h_stream),
            len(d_stream),
            len(q_stream),
            len(r_stream)
        )
        
        return header_bytes + h_stream + d_stream + q_stream + r_stream + struct.pack("<I", crc)

    def decompress_block(self, in_stream: BinaryIO) -> List[FASTQRecord]:
        """Decompresses an optimized .gqz block."""
        header_size = struct.calcsize("<4sIIHBIIII")
        header_raw = in_stream.read(header_size)
        if not header_raw:
            return []
        if len(header_raw) < header_size:
            raise ValueError(f"Truncated container block header: expected {header_size} bytes, got {len(header_raw)} bytes.")
            
        marker, record_count, uncompressed_size, flags, b_level, h_size, d_size, q_size, r_size = struct.unpack(
            "<4sIIHBIIII", header_raw
        )
        if marker != BLOCK_MARKER:
            raise ValueError(f"Corrupted container marker: expected {BLOCK_MARKER!r}, got {marker!r}.")
        if record_count == 0:
            return []
            
        h_stream = in_stream.read(h_size)
        d_stream = in_stream.read(d_size)
        q_stream = in_stream.read(q_size)
        r_stream = in_stream.read(r_size) if r_size > 0 else b""
        crc_raw = in_stream.read(4)
        stored_crc = struct.unpack("<I", crc_raw)[0]
        
        # Checksum check
        crc = zlib.crc32(h_stream)
        crc = zlib.crc32(d_stream, crc)
        crc = zlib.crc32(q_stream, crc)
        if r_stream:
            crc = zlib.crc32(r_stream, crc)
        if (crc & 0xFFFFFFFF) != (stored_crc & 0xFFFFFFFF):
            raise ValueError(f"Block CRC32 checksum mismatch: calculated {crc:#010x} vs stored {stored_crc:#010x}")
            
        # Decode components
        headers = self._decode_headers(h_stream, record_count)
        dna_packed = RansCodec.decompress_bytes(d_stream)
        sequences = self._unpack_dna(dna_packed)
        quant_quals = self._decode_quality_stream(q_stream, record_count)
        
        raw_res_bytes = RansCodec.decompress_bytes(r_stream) if (flags & 0x0001 and r_stream) else b""
        residuals = [zigzag_decode(b) for b in raw_res_bytes] if raw_res_bytes else []
        
        records = []
        res_offset = 0
        for i in range(record_count):
            h = headers[i] if i < len(headers) else f"@READ_{i}"
            s = sequences[i] if i < len(sequences) else ""
            q_line = quant_quals[i] if i < len(quant_quals) else ""
            
            if flags & 0x0001 and residuals and res_offset + len(q_line) <= len(residuals):
                sub_res = residuals[res_offset:res_offset + len(q_line)]
                final_q = Quantizer.restore(q_line, sub_res)
                res_offset += len(q_line)
            else:
                final_q = q_line
            records.append(FASTQRecord(header=h, sequence=s, quality=final_q))
            
        return records

    def compress_stream(self, in_file: BinaryIO, out_file: BinaryIO) -> CompressionStats:
        """Compresses binary FASTQ stream with strict memory consumption checks (<4 GB)."""
        t_start = time.perf_counter()
        stats = CompressionStats()
        process = psutil.Process(os.getpid())
        
        flags = 1 if self.options.lossless else 0
        out_file.write(MAGIC_HEADER)
        out_file.write(struct.pack("<HBB", FORMAT_VERSION, int(self.options.binning_level), flags))
        
        from gqzip.license import check_allowance_permission, update_processed_bytes
        ok, msg = check_allowance_permission()
        if not ok:
            raise PermissionError(msg)

        batch: List[FASTQRecord] = []
        
        while True:
            mem_mb = process.memory_info().rss / (1024.0 * 1024.0)
            if mem_mb > stats.peak_memory_mb:
                stats.peak_memory_mb = mem_mb
            assert mem_mb < 4000.0, f"Memory safety violation: Peak RSS {mem_mb:.2f} MB exceeded 4.0 GB!"
            
            h_line = in_file.readline()
            if not h_line:
                break
            h_str = h_line.decode('latin1').strip()
            if not h_str:
                continue
                
            s_str = in_file.readline().decode('latin1').strip()
            p_str = in_file.readline().decode('latin1').strip()
            q_str = in_file.readline().decode('latin1').strip()

            if not s_str or not q_str:
                # Incomplete partial FASTQ record at EOF; skip
                break
                
            if len(s_str) != len(q_str):
                min_len = min(len(s_str), len(q_str))
                sys.stderr.write(
                    f"\n[GQZIP WARNING] Read length mismatch in record '{h_str}': "
                    f"sequence length ({len(s_str)}) != quality length ({len(q_str)}). Auto-trimmed to {min_len} bp.\n"
                )
                s_str = s_str[:min_len]
                q_str = q_str[:min_len]
                
            if not h_str.startswith('@'):
                h_str = '@' + h_str
                
            batch.append(FASTQRecord(header=h_str, sequence=s_str, quality=q_str))
            stats.raw_fastq_bytes += len(h_str) + len(s_str) + len(q_str) + 4
            stats.raw_quality_bytes += len(q_str)
            stats.total_records += 1
            
            if len(batch) >= self.options.block_records:
                block_bytes = self.compress_batch(batch)
                out_file.write(block_bytes)
                stats.compressed_bytes += len(block_bytes)
                batch.clear()
                
        if batch:
            block_bytes = self.compress_batch(batch)
            out_file.write(block_bytes)
            stats.compressed_bytes += len(block_bytes)
            batch.clear()

        update_processed_bytes(stats.raw_fastq_bytes)
            
        # End of stream
        out_file.write(struct.pack("<4sIIHBIIII", BLOCK_MARKER, 0, 0, 0, 0, 0, 0, 0, 0))
        
        t_end = time.perf_counter()
        stats.compression_time_sec = t_end - t_start
        return stats

    def decompress_stream(self, in_file: BinaryIO, out_file: BinaryIO) -> CompressionStats:
        """Decompresses binary .gqz stream to standard 4-line FASTQ."""
        t_start = time.perf_counter()
        stats = CompressionStats()
        
        magic = in_file.read(4)
        if magic != MAGIC_HEADER:
            raise ValueError(f"Invalid gqzip header: expected {MAGIC_HEADER!r}, got {magic!r}")
            
        ver, b_level, flags = struct.unpack("<HBB", in_file.read(4))
        
        while True:
            records = self.decompress_block(in_file)
            if not records:
                break
                
            for r in records:
                out_file.write(f"{r.header}\n{r.sequence}\n+\n{r.quality}\n".encode('latin1'))
                stats.raw_fastq_bytes += len(r.header) + len(r.sequence) + len(r.quality) + 4
                stats.total_records += 1
                
        t_end = time.perf_counter()
        stats.decompression_time_sec = t_end - t_start
        return stats

def compress_file(input_path: str, output_path: str, options: CompressionOptions = None) -> CompressionStats:
    engine = CompressionEngine(options)
    with open(input_path, 'rb') as fin, open(output_path, 'wb') as fout:
        return engine.compress_stream(fin, fout)

def decompress_file(input_path: str, output_path: str, options: CompressionOptions = None) -> CompressionStats:
    engine = CompressionEngine(options)
    with open(input_path, 'rb') as fin, open(output_path, 'wb') as fout:
        return engine.decompress_stream(fin, fout)
